python/sglang/srt/server.py

```diff
@@ -123,31 +123,91 @@ async def flush_cache():
+    """
+    # NOTE: This is because the multiple requests in one http request.
+
+    async def convert_style(r, return_text):
+        r["meta_info"]["prefill_token_logprobs"] = await detokenize_logprob_tokens(
+            r["meta_info"]["prefill_token_logprobs"], return_text
+        )
+        r["meta_info"]["decode_token_logprobs"] = await detokenize_logprob_tokens(
+            r["meta_info"]["decode_token_logprobs"], return_text
+        )
+        r["meta_info"]["prefill_top_logprobs"] = await detokenize_top_logprobs_tokens(
+            r["meta_info"]["prefill_top_logprobs"], return_text
+        )
+        r["meta_info"]["decode_top_logprobs"] = await detokenize_top_logprobs_tokens(
+            r["meta_info"]["decode_top_logprobs"], return_text
+        )
+
+    if isinstance(obj.text, str):
+        if obj.return_logprob:
+            await convert_style(ret, obj.return_text_in_logprobs)
+    else:
+        for i, r in enumerate(ret):
+            if obj.return_logprob[i]:
+                await convert_style(r, obj.return_text_in_logprobs)
 
 
 async def stream_generator(obj: GenerateReqInput):
     async for out in tokenizer_manager.generate_request(obj):
-        if obj.return_logprob and obj.return_text_in_logprobs:
-            out["meta_info"]["token_logprob"] = await detokenize_logprob_tokens(
-                out["meta_info"]["token_logprob"]
-            )
+        await handle_token_logprobs_results(obj, out)
         yield out
 
 
-async def make_openai_style_logprobs(token_logprobs):
+async def make_openai_style_logprobs(
+    prefill_token_logprobs=None,
+    decode_token_logprobs=None,
+    prefill_top_logprobs=None,
+    decode_top_logprobs=None,
+):
     ret_logprobs = LogProbs()
 
-    for token_text, token_logprob in token_logprobs:
-        ret_logprobs.tokens.append(token_text)
-        ret_logprobs.token_logprobs.append(token_logprob)
+    def append_token_logprobs(token_logprobs):
+        for logprob, _, token_text in token_logprobs:
+            ret_logprobs.tokens.append(token_text)
+            ret_logprobs.token_logprobs.append(logprob)
+
+    def append_top_logprobs(top_logprobs):
+        for tokens in top_logprobs:
+            if tokens is not None:
+                ret_logprobs.top_logprobs.append(
+                    {token[2]: token[0] for token in tokens}
+                )
+            else:
+                ret_logprobs.top_logprobs.append(None)
+
+    if prefill_token_logprobs is not None:
+        append_token_logprobs(prefill_token_logprobs)
+    if decode_token_logprobs is not None:
+        append_token_logprobs(decode_token_logprobs)
+    if prefill_top_logprobs is not None:
+        append_top_logprobs(prefill_top_logprobs)
+    if decode_top_logprobs is not None:
+        append_top_logprobs(decode_top_logprobs)
 
-        # Not supported yet.
-        ret_logprobs.top_logprobs.append({})
-        ret_logprobs.text_offset.append(-1)
```

Comment:
Should we keep this?

Reply:
We now support `top_logprobs`.
The `text_offset` is meaningless as we return both prefilling tokens logprobs and decoding tokens' logprobs.

Reply:
Since this is a field in OpenAI SDK, I'm not sure if missing this would introduce incompatibility in someone's integration. Maybe we could just leave a reminder.

Reply:
https://github.com/sgl-project/sglang/pull/331/commits/cfa7e65c62424baf7634f23f44cc05c4e551d829

I fixed it in this